### PR TITLE
Closes #156

### LINKED ISSUE #156: Create dataset loader for MSH-WSD dataset
- **Name:** MSH-WSD Dataset
- **Description:** 
- **Task:** Word Sense Disambiguation
- **Paper:** https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-223
- **Data:** https://lhncbc.nlm.nih.gov/ii/areas/WSD/collaboration.html
- **License:** ?
- **Motivation:** completeness. but this one looks hard.

### PR DESCRIPTION
**Pull request by nomisto**
Closes #156

Local dataset, output in file: [test_log.txt](https://github.com/bigscience-workshop/biomedical/files/8473859/myoutput.txt)

Dataset is of task Word Sense Disambiguation fitted into Tasks.NAMED_ENTITY_DISAMBIGUATION. There was a discussion starting from [here](https://discord.com/channels/945077536260497479/945077620666667029/963356609273266186).

There are a lot of mismatches between entity and offset, this is intended as there can be slight variations of ambiguous word/word in text (in case, singular/plural, ...).

### CODE REVIEW

**Comment by nomisto** [comment]
This is my first local dataset, when I run `a = datasets.load_dataset(f"./biodatasets/msh_wsd/msh_wsd.py", data_dir=r"C:\Users\ottsi\Downloads\MSHCorpus\MSHCorpus")`, so that it should select the default config it throws the follwoing error. I can't figure out the cause. I know this sounds weird, but is it possible that this has something to do with it being local? 

Output of error:
```bash
Traceback (most recent call last):
  File "c:\Users\ottsi\biomedical\test.py", line 10, in <module>
    a = datasets.load_dataset(f"./biodatasets/{sys.argv[1]}/{sys.argv[1]}.py", data_dir=r"C:\Users\ottsi\Downloads\MSHCorpus")
  File "c:\Users\ottsi\biomedical\venv\lib\site-packages\datasets\load.py", line 1675, in load_dataset
    builder_instance = load_dataset_builder(
  File "c:\Users\ottsi\biomedical\venv\lib\site-packages\datasets\load.py", line 1533, in load_dataset_builder
    builder_instance: DatasetBuilder = builder_cls(
  File "c:\Users\ottsi\biomedical\venv\lib\site-packages\datasets\builder.py", line 1020, in __init__
    super().__init__(*args, **kwargs)
  File "c:\Users\ottsi\biomedical\venv\lib\site-packages\datasets\builder.py", line 268, in __init__
    info.update(self._info())
  File "C:\Users\ottsi\.cache\huggingface\modules\datasets_modules\datasets\msh_wsd\ec9960ad15241a249d9f3bc0747fe8c89b4e172f0c9012bc71b0209004840c39\msh_wsd.py", line 112, in _info
    if self.config.schema == "source":
AttributeError: 'BuilderConfig' object has no attribute 'schema'
```

When I print `self.config` in L112 i get the following
```bash
BuilderConfig(name='default', version=0.0.0, data_dir='C:\\Users\\ottsi\\Downloads\\MSHCorpus\\MSHCorpus', data_files=None, description=None)
```

When I print `self.config` in L112 without `data_dir` it selects the correct default config:
```bash
BigBioConfig(name='msh_wsd_source', version=1.0.0, data_dir=None, data_files=None, description='MSH-WSD source schema', schema='source', subset_id='msh_wsd')
```

File: biodatasets/msh_wsd/msh_wsd.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+            document_["coreferences"] = []
+            document_["id"] = next(self.uid)
+            document_["document_id"] = sentence["pmid"]
+            document_["passages"] = [
+                {"id": next(self.uid), "type": "", "text": [sentence["text"]], "offsets": [[0, len(sentence["text"])]]}
+            ]
+            document_["entities"] = [
+                {
+                    "id": next(self.uid),
+                    "type": "",
```
Comment (sg-wbi) [inline biodatasets/msh_wsd/msh_wsd.py:236]: How about "ambigous_word"?

**Comment by sg-wbi** [comment]
Thank you adding this dataset @nomisto ! Just to be on the safe side, this commit [827f2b9](https://github.com/bigscience-workshop/biomedical/pull/434/commits/827f2b963372883dc80e41e75c6f68eeb6521e57) resolve the issue you had with the testing script?

File: biodatasets/msh_wsd/msh_wsd.py (hunk 1 above)
Comment (nomisto) [reply biodatasets/msh_wsd/msh_wsd.py:236]: Sorry, could you clarify? I don't understand what you mean unfortunately

**Comment by nomisto** [comment]
No problem! Unfortunately not @sg-wbi, I have this problem with another local dataset I'm currently working on, so this might be a bug in `load_datasets`?

**Comment by nomisto** [comment]
Last two commits are due to @galtay s information that `data_dir` should point to the zip file rather than the unzipped data.

File: biodatasets/msh_wsd/msh_wsd.py (hunk 1 above)
Comment (sg-wbi) [reply biodatasets/msh_wsd/msh_wsd.py:235]: Just a minor thing `"type": "ambiguous_word"`

Comment (nomisto) [reply biodatasets/msh_wsd/msh_wsd.py:235]: Ah got you! Thanks!

File: biodatasets/msh_wsd/msh_wsd.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        ),
+        BigBioConfig(
+            name="msh_wsd_bigbio_kb",
+            version=BIGBIO_VERSION,
+            description="MSH-WSD BigBio schema",
+            schema="bigbio_kb",
+            subset_id="msh_wsd",
+        ),
+    ]
+
```
Comment (sg-wbi) [inline biodatasets/msh_wsd/msh_wsd.py:114]: If I understood correctly from #364 you need to add this: 
```python
BUILDER_CONFIG_CLASS = BigBioConfig
```

**Comment by sg-wbi** [comment]
But you still need to pass the config `name` when you load the dataset for it to work
```bash
datasets.load_dataset(f"./biodatasets/msh_wsd/msh_wsd.py", data_dir=r"C:\Users\ottsi\Downloads\MSHCorpus\MSHCorpus", "msh_wsd_source")
```

**Comment by sg-wbi** [comment]
For reference: #454

**Comment by galtay** [comment]
hi! going to try a little PR surgery here.
* sorry about the confusion with data_dir ... I think it should point to a directory that contains `MSHCorpus.zip` and I'll implement that 
* I think for now we'll go with the subclass of BigBioConfig for the other set of errors (as was done in #364

**Comment by galtay** [comment]
I'm getting some decoding errors when reading some of these `arff` files ... have you encountered that? 
one of them is `MSHCorpus/Erythrocytes_pmids_tagged.arff`

**Comment by galtay** [comment]
> hi! going to try a little PR surgery here.
> 
>     * sorry about the confusion with data_dir ... I think it should point to a directory that contains `MSHCorpus.zip` and I'll implement that
> 
>     * I think for now we'll go with the subclass of BigBioConfig for the other set of errors (as was done in [Closes #246 #364](https://github.com/bigscience-workshop/biomedical/pull/364)

OK new plan :) I'll put some comments in the PR ... don't want to make changes until I hear back about the file decoding to see if it's just on my end.

File: biodatasets/msh_wsd/msh_wsd.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+System (UMLS) Metathesaurus and the manual MeSH indexing of MEDLINE. The resulting dataset is called MSH WSD and
+consists of 106 ambiguous abbreviations, 88 ambiguous terms and 9 which are a combination of both, for a total of 203
+ambiguous words. Each instance containing the ambiguous word was assigned a CUI from the 2009AB version of the UMLS.
+For each ambiguous term/abbreviation, the data set contains a maximum of 100 instances per sense obtained from
+MEDLINE; totaling 37,888 ambiguity cases in 37,090 MEDLINE citations.
+
+Note from the Author how to load dataset:
+1) Download the file MSHCorpus.zip (Link "MSHWSD Data Set") from
+   https://lhncbc.nlm.nih.gov/ii/areas/WSD/collaboration.html
+2) Set kwarg data_dir to the path of the zip file
```
Comment (galtay) [inline biodatasets/msh_wsd/msh_wsd.py:29]: 2) Set kwarg data_dir to the directory containing MSHCorpus.zip

File: biodatasets/msh_wsd/msh_wsd.py (hunk 2 above)
Comment (galtay) [reply biodatasets/msh_wsd/msh_wsd.py:114]: yes, for now I think we can use the solution from #364 with a subclass of `BigBioConfig` defined here with the defaults that you want and `BUILDER_CONFIG_CLASS` set to that class you define.

File: biodatasets/msh_wsd/msh_wsd.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager) -> List[datasets.SplitGenerator]:
+        """Returns SplitGenerators."""
+
+        if self.config.data_dir is None:
+            raise ValueError("This is a local dataset. Please pass the data_dir kwarg to load_dataset.")
+        else:
+            data_dir = data_dir = dl_manager.download_and_extract(self.config.data_dir)
```
Comment (galtay) [inline biodatasets/msh_wsd/msh_wsd.py:154]: `data_dir = dl_manager.download_and_extract(os.path.join(self.config.data_dir, "MSHCorpus.zip"))`

File: biodatasets/msh_wsd/msh_wsd.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+                    yield guid, example
+
+            elif self.config.schema == "bigbio_kb":
+                for document in self._parse_document(concept_map, file):
+                    for example in self._source_to_kb(document):
+                        yield example["id"], example
+
+    def _parse_document(self, concept_map, file: Path):
+
+        with file.open() as f:
```
Comment (galtay) [inline biodatasets/msh_wsd/msh_wsd.py:188]: can we open in read or read binary model (depending on what happens with the decoding)?

**Review by galtay — CHANGES_REQUESTED** [review]

**Comment by nomisto** [comment]
Hi Gabriel, thank's for your review and chatching the encoding error. I've implemented and tested the script using Windows which uses a different default encoding. I have written the correct encoding in the comment above (and also tested it using linux).

File: biodatasets/msh_wsd/msh_wsd.py (hunk 5 above)
Comment (nomisto) [reply biodatasets/msh_wsd/msh_wsd.py:188]: Replacing L188 with `with file.open(mode="r", encoding="iso-8859-1") as f:` should fix this.

**Comment by galtay** [comment]
> Hi Gabriel, thank's for your review and chatching the encoding error. I've implemented and tested the script using Windows which uses a different default encoding. I have written the correct encoding in the comment above (and also tested it using linux).

good to know! I initially thought I could do the PR surgery myself, but the method I wanted to use for merging didn't work. can you make the updates please?

**Comment by nomisto** [comment]
Sure! 43c2667 should include all changes. The output of the tests remain the same.

**Comment by sg-wbi** [comment]
Hey @galtay just checking: I've seen you merged #447 which is - to my eyes - essentially identical to this one. Is there something else blocking or can we merge this? Thanks

**Comment by sg-wbi** [comment]
@galtay I assume you left me the honor of merging so I;ll go ahead with this one. Maybe I;ll get a banner saying "you;ve just merged the 100th dataset" :1st_place_medal: 
Thanks again for the contribution @nomisto !

**Pull request merged**